From 5bfd44849ffd6d1eac4eaa0d23b547a545de0194 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Thu, 2 Oct 2014 14:24:18 -0400
Subject: [fs] nfsd4: fix corruption of NFSv4 read data

Message-id: <1412259858-791-1-git-send-email-bfields@redhat.com>
Patchwork-id: 96881
O-Subject: [Patch RHEL7 fs] nfsd4: fix corruption of NFSv4 read data
Bugzilla: 1148055
RH-Acked-by: Andrea Arcangeli <aarcange@redhat.com>
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>

The calculation of page_ptr here is wrong in the case the read doesn't
start at an offset that is a multiple of a page.

The result is that nfs4svc_encode_compoundres sets rq_next_page to a
value one too small, and then the loop in svc_free_res_pages may
incorrectly fail to clear a page pointer in rq_respages[].

Pages left in rq_respages[] are available for the next rpc request to
use, so xdr data may be written to that page, which may hold data still
waiting to be transmitted to the client or data in the page cache.

The observed result was silent data corruption seen on an NFSv4 client.

We tag this as "fixing" 05638dc73af2 because that commit exposed this
bug, though the incorrect calculation predates it.

Particular thanks to Andrea Arcangeli and David Gilbert for analysis and
testing.

Fixes: 05638dc73af2 "nfsd4: simplify server xdr->next_page use"
Cc: stable@vger.kernel.org
Reported-by: Andrea Arcangeli <aarcange@redhat.com>
Tested-by: "Dr. David Alan Gilbert" <dgilbert@redhat.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>

Upstream: 15b23ef5d348ea51c5e7573e2ef4116fbc7cb099
Bugzilla: 1148055
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4xdr.c b/fs/nfsd/nfs4xdr.c
index e94457c..b01f6e1 100644
--- a/fs/nfsd/nfs4xdr.c
+++ b/fs/nfsd/nfs4xdr.c
@@ -3104,7 +3104,8 @@ static __be32 nfsd4_encode_splice_read(
 
  buf->page_len = maxcount;
  buf->len += maxcount;
- xdr->page_ptr += (maxcount + PAGE_SIZE - 1) / PAGE_SIZE;
+ xdr->page_ptr += (buf->page_base + maxcount + PAGE_SIZE - 1)
+       / PAGE_SIZE;
 
  /* Use rest of head for padding and remaining ops: */
  buf->tail[0].iov_base = xdr->p;
-- 
1.7.1